storage/google_drive/drive_client_wrapper.py

```diff
@@ -214,24 +237,38 @@ def update_or_create(source_file_path, target_folder_path, target_file_name=None
             if existing_file.get("mimetype") == "application/vnd.google-apps.folder":
                 log.error(f"Attempting to replace a folder with a file with name '{target_file_name}'")
                 exit(1)
-            _update_file(source_file_path, existing_file.get("id"))
-            return
+            _auto_retry(lambda: _update_file(source_file_path, existing_file.get("id")), max_retries, backoff_seconds)
+            continue
 
-        _create_file(source_file_path, target_folder_id, target_file_name)
-    except (HttpError, socket.timeout) as ex:
-        if type(ex) == HttpError:
-            if ex.resp.status not in {500, 503}:
-                raise ex
-            log.warning(f"Upload failed with HttpError {ex.resp.status}")
+        _auto_retry(lambda: _create_file(source_file_path, target_folder_id, target_file_name),
+                    max_retries, backoff_seconds)
 
-        if type(ex) == socket.timeout:
-            log.warning(f"Upload failed with socket.timeout error")
 
-        if max_retries > 0:
-            log.info(f"Retrying up to {max_retries} more times, after {backoff_seconds} seconds...")
-            time.sleep(backoff_seconds)
-            update_or_create(source_file_path, target_folder_path, target_file_name, recursive,
-                             target_folder_is_shared_with_me, max_retries - 1, backoff_seconds * 2)
-        else:
-            log.error("Retried the maximum number of times")
-            raise ex
+def update_or_create(source_file_path, target_folder_path, target_file_name=None, recursive=False,
+                     target_folder_is_shared_with_me=False, max_retries=2, backoff_seconds=1):
+    if target_file_name is None:
+        target_file_name = os.path.basename(source_file_path)
+
+    target_folder_id = _auto_retry(
+        lambda: _get_path_id(target_folder_path, recursive, target_folder_is_shared_with_me),
+        max_retries, backoff_seconds)
+    files = _auto_retry(lambda: _list_folder_id(target_folder_id), max_retries, backoff_seconds)
+
+    files_with_upload_name = list(filter(lambda file: file.get('name') == target_file_name, files))
+
+    if len(files_with_upload_name) > 1:
+        log.error("Multiple files with the same name found in Drive folder.")
+        log.error("I don't know which to update, aborting.")
+        exit(1)
+
+    if len(files_with_upload_name) == 1:
+        existing_file = files_with_upload_name[0]
+        # Make sure it's not a folder
+        if existing_file.get("mimetype") == "application/vnd.google-apps.folder":
```

Comment:
We should probably define a 'constant' for this to reduce errors during maintenance.

Reply:
Moved to DRIVE_FOLDER_TYPE